### PR TITLE
Initializes the restrictions list for each network

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

Initializes the restrictions list for each network.

This PR introduces the concept of restrictions for `programs`, `functions`, and `arguments`.
- Allows for a network to restrict usage of an entire program from all transitions
- Allows for a network to restrict usage of any specified program function from all transitions
- Allows for a network to restrict usage of any plaintext literal input or output for any program function call from all transitions

Currently, the `mainnet`, `testnet`, and `canary` networks are initialized to an empty restrictions list, meaning there are no restrictions at inception. It is my understanding that a governance process will be overseen with the community to update any of these restrictions as necessary over time.

Please be aware that snarkOS in PR https://github.com/ProvableHQ/snarkOS/pull/14 enforces in its handshakes that all nodes are on matching `restrictions ID` in order to connect to one another over the network.

### CODE REVIEW

**Review by vicsn — COMMENTED** [review]
To help future readers, can you replace the PR message:
- from: 'Allows for a network to restrict usage of any input or output for any program function call from all transitions'
- to: 'Allows for a network to restrict usage of plaintext Literal input or output for any program function call from all transitions'

> Please be aware that snarkOS will enforce in its handshakes that all nodes are on matching restrictions ID in order to connect to one another over the network.

Good idea. Can you explain where this logic is or will be implemented?

File: synthesizer/src/restrictions/helpers/argument_locator.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    #[test]
+    fn test_import_parse() {
+        let rng = &mut TestRng::default();
+
+        for _ in 0..ITERATIONS {
+            let expected = sample_argument_locator(rng);
+
+            // Test the parser.
+            let arugment_locator = ArgumentLocator::parse(&expected.to_string()).unwrap().1;
+            assert_eq!(expected.is_input(), arugment_locator.is_input());
```
[the remark this answers is not in the document]
Comment (howardwu) [reply synthesizer/src/restrictions/helpers/argument_locator.rs]: Resolved in [e579d0a](https://github.com/ProvableHQ/snarkVM/pull/15/commits/e579d0a9f30ec3aa555674652d0f77bffe4c0a74)

File: synthesizer/src/restrictions/mod.rs [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    /// e.g. `restricted.aleo/bar _ aleo1zkpxxxxx _ _` => `10..` (from block 10 onwards)
+    /// e.g. `restricted.aleo/bar _ aleo1zkpxxxxx _ _` => `..10` (up to block 10)
+    /// e.g. `restricted.aleo/bar _ aleo1zkpxxxxx _ _` => `10..20` (from block 10 to block 20)
+    ///
+    /// Note: This design intentionally minimizes the number of total lookups required to check
+    /// for restrictions when a transition matches the described profile. In summary:
+    /// - When a transition does not match the program ID or function name, the total lookup cost is `O(1)`.
+    /// - When a transition matches the program ID & function name, the initial lookup cost is `O(num_inputs + num_outputs)`.
+    ///    - If an input or output index does not match, the additional lookup cost is `0`.
+    ///    - If an input or output index matches, the additional lookup cost is `O(n)` for `n` arguments with the same index.
```
[the remark this answers is not in the document]
Comment (howardwu) [reply synthesizer/src/restrictions/mod.rs]: Resolved in [e579d0a](https://github.com/ProvableHQ/snarkVM/pull/15/commits/e579d0a9f30ec3aa555674652d0f77bffe4c0a74)

**Comment by vicsn** [comment]
Closing here as it is now opened against AleoNet

**Comment by damons** [comment]
What's an adequate test plan for this feature?  Need to know the expected behavior when blocking programs, functions, and arguments.

**Comment by howardwu** [comment]
Agreed, this is an important discussion. Can we have a discussion with @apruden2008 on the matter?

For background context, the initial ask was to have the restrictions list feature implemented in a lightweight and prompt fashion to reflect the realities of the timeline for code freeze and relative priorities on the table.

**Pull request closed without merging**